pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,522 @@
+	ipPoolLister            crdlister.ExternalIPPoolLister
+	ipPoolInformerHasSynced cache.InformerSynced
+}
+
+// NewCluster return a new *Cluster
+func NewCluster(
+	clusterBindPort int,
+	nodeInformer coreinformers.NodeInformer,
+	nodeConfig *config.NodeConfig,
+	ipPoolInformer crdinformers.ExternalIPPoolInformer,
+) (*Cluster, error) {
+	// The Node join/leave events will be notified via it.
+	nodeEventCh := make(chan memberlist.NodeEvent, 1024)
+	c := &Cluster{
+		bindPort:                clusterBindPort,
+		NodeConfig:              nodeConfig,
+		nodeName:                nodeConfig.Name,
+		nodeEventsCh:            nodeEventCh,
+		nodeInformer:            nodeInformer,
+		nodeLister:              nodeInformer.Lister(),
+		nodeListerSynced:        nodeInformer.Informer().HasSynced,
+		ipPoolInformer:          ipPoolInformer,
+		ipPoolLister:            ipPoolInformer.Lister(),
+		ipPoolInformerHasSynced: ipPoolInformer.Informer().HasSynced,
+		localNodeEventCh:        make(chan bool, 1),
+	}
+	c.conHash = newNodeConsistentHashMap()
+
+	bindPort := c.bindPort
+	hostIP := c.NodeConfig.NodeIPAddr.IP
+	nodeMember := fmt.Sprintf("%s:%d", hostIP.String(), bindPort)
+	klog.V(2).InfoS("Add new node", "local node", nodeMember)
+
+	conf := memberlist.DefaultLocalConfig()
+	conf.Name = c.nodeName
+	conf.BindPort = bindPort
+	conf.AdvertisePort = bindPort
+	conf.Events = &memberlist.ChannelEventDelegate{Ch: nodeEventCh}
+	klog.V(1).Infof("Memberlist cluster configs: %+v", conf)
+
+	mList, err := memberlist.Create(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create memberlist: %v", err)
+	}
+
+	c.mlConfig = conf
+	c.mList = mList
+	c.existingMembers = sets.NewString(nodeMember)
+
+	nodeInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				node, ok := obj.(*corev1.Node)
+				if !ok {
+					klog.ErrorS(unexpectError, "Add node callback", "obj", obj)
+					return
+				}
+				klog.V(2).InfoS("Add node event", "nodeName", node.Name)
+				shouldJoin := c.shouldJoinCluster(node.Name)
+				if shouldJoin {
+					c.updateExistingMember(node, true)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				node, ok := newObj.(*corev1.Node)
+				oldNode, okOldNode := oldObj.(*corev1.Node)
+				if !ok || !okOldNode {
+					klog.ErrorS(unexpectError, "Update node callback", "oldObj", oldObj, "newObj", newObj)
+					return
+				}
+				if reflect.DeepEqual(node.GetLabels(), oldNode.GetLabels()) {
+					klog.V(2).InfoS("Update node event, labels not changed", "nodeName", node.Name)
+					return
+				}
+				oldMatch, newMatch := c.matchNodeSelectors(oldNode), c.matchNodeSelectors(node)
+				if oldMatch == newMatch {
+					klog.V(2).InfoS("Update node event, node cluster status not changed", "nodeName", node.Name)
+					return
+				}
+				klog.V(2).InfoS("Update node event", "nodeName", node.Name)
+				c.updateExistingMember(node, newMatch && !oldMatch)
```

Comment:
At the moment newMatch already equals to !oldMatch. It can just use `newMatch`?

Reply:
done